src/Jobs/RemoveDataObjectJob.php

```diff
@@ -52,20 +52,36 @@ public function setup(): void
         $documents = Versioned::withVersionedMode(function () use ($archiveDate) {
             Versioned::reading_archived_date($archiveDate);
 
+            $currentDocument = $this->getDocument();
             // Go back in time to find out what the owners were before unpublish
-            $dependentDocs = $this->document->getDependentDocuments();
+            $dependentDocs = $currentDocument->getDependentDocuments();
 
             // refetch everything on the live stage
             Versioned::set_stage(Versioned::LIVE);
 
-            return array_map(function (DataObjectDocument $doc) {
-                return DataObjectDocument::create(
-                    DataObject::get_by_id(
-                        $doc->getSourceClass(),
-                        $doc->getDataObject()->ID
-                    )
-                );
-            }, $dependentDocs);
+            return array_reduce(
+                $dependentDocs,
+                function (array $carry, DataObjectDocument $doc) {
+                    $record = DataObject::get_by_id($doc->getSourceClass(), $doc->getDataObject()->ID);
+
+                    if ($record) {
+                        $document = DataObjectDocument::create($record);
+                        $carry[$document->getIdentifier()] = $document;
+
+                        return $carry;
+                    }
+
+                    $oldRecord = $doc->getDataObject();
+
+                    if ($oldRecord->isArchived() || $oldRecord->isOnDraft()) {
+                        $document = DataObjectDocument::create($oldRecord);
+                        $carry[$document->getIdentifier()] = $document;
+                    }
+
+                    return $carry;
+                },
+                []
+            );
```

Comment:
It might just be because I'm not familiar with this module, but it's not particularly clear to me what's happening here. It might pay to add a comment about this chunk of code with a brief explanation of what's going on.

Reply:
Have added comments as requested @GuySartorelli.